compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -383,6 +383,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let mut oprnd_t = self.check_expr_with_expectation(&oprnd, expected_inner);
 
         if !oprnd_t.references_error() {
+            match (unop, oprnd_t.kind()) {
```

Comment:
This inference strategy is incorrect. There's no guarantee that the input and output type of `Neg` or `Not` are related, and this causes this ICE in the compiler:

```

#[derive(Copy, Clone, Default)]
struct A;

struct B;

impl std::ops::Not for A {
    type Output = B;

    fn not(self) -> B { B }
}

fn main() {
    let x = Default::default();
    let y: A = !x;
}
```

Reply:
In this example `x` is constrained to be `A` because of the inference, but that would imply that `y` is type `B`, not `A`.

Reply:
make sense 👍
this is a trivial difference here, `binary op` is more eager, so `unary` will infer type error.
 
```rust
pub fn bools(x: &Vec<bool>) {
    let binary = |i, a: &Vec<bool>| {
        a[i] && a[i+1] // ok
    };

    let unary = |i, a: &Vec<bool>| {
        !a[i] // cannot infer type
    };

    binary(0, x);
    unary(0, x);
}
```

Seems we can not find proper fix for this scenario?